### PR TITLE
contrib/internal/httptrace: improve performance of span starts

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This commit improves the performance of starting a span by reducing work
done, especially allocations.

### Motivation

We were seeing a large increase in memory and CPU usage while starting spans compared to previous versions.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

**Comment by knusbaum** [comment]
```
$ benchstat old.bench new.bench 
name                 old time/op    new time/op    delta
StartRequestSpan-16     788ns ±13%     160ns ± 6%  -79.64%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
StartRequestSpan-16      768B ± 0%      224B ± 0%  -70.83%  (p=0.000 n=10+10)

name                 old allocs/op  new allocs/op  delta
StartRequestSpan-16      21.0 ± 0%       5.0 ± 0%  -76.19%  (p=0.000 n=10+10)
```

File: internal/namingschema/op_client_server.go [hunk 1]
```diff
@@ -37,13 +37,32 @@ type serverInboundOp struct {
...
+
+func (s SimpleSchema) Name() string { return s() }
+
+type Namer interface {
+	Name() string
+}
+
 // NewServerInboundOp creates a new naming schema for server inbound operations.
-func NewServerInboundOp(system string, opts ...Option) *Schema {
+func NewServerInboundOp(system string, opts ...Option) Namer {
```
Comment (knusbaum) [inline internal/namingschema/op_client_server.go:49]: This is more or less a hack. It's not consistent with the rest of the namingschema package.

We should refactor the whole package to do something similar to this.

File: contrib/internal/httptrace/httptrace.go [hunk 2]
```diff
@@ -23,36 +23,50 @@ import (
...
 	if cfg.traceClientIP {
-		ipTags, _ := httpsec.ClientIPTags(r.Header, true, r.RemoteAddr)
-		for k, v := range ipTags {
-			opts = append(opts, tracer.Tag(k, v))
-		}
+		ipTags, _ = httpsec.ClientIPTags(r.Header, true, r.RemoteAddr)
 	}
-	return tracer.StartSpanFromContext(r.Context(), namingschema.NewHTTPServerOp().GetName(), opts...)
+	nopts := make([]ddtrace.StartSpanOption, 0, len(opts)+7+len(ipTags))
+	// nopts = append(nopts,
```
Comment (knusbaum) [inline contrib/internal/httptrace/httptrace.go:41]: TODO: dead code

File: contrib/internal/httptrace/httptrace_test.go [hunk 3]
```diff
@@ -237,3 +239,20 @@ func TestURLTag(t *testing.T) {
...
+	r, err := http.NewRequest("GET", "http://example.com", nil)
+	if err != nil {
+		b.Errorf("Failed to create request: %v", err)
+		return
+	}
+	opts := []ddtrace.StartSpanOption{
+		tracer.ServiceName("SomeService"),
+		tracer.ResourceName("SomeResource"),
+		tracer.Tag(ext.HTTPRoute, "/some/route/?"),
+	}
```
Comment (ajgajg1134) [inline contrib/internal/httptrace/httptrace_test.go:254]:
```suggestion
	}
	b.ResetTimer()
```

Comment (knusbaum) [reply contrib/internal/httptrace/httptrace_test.go:254]: :+1:

**Review by ajgajg1134 — APPROVED** [review]

**Comment by knusbaum** [comment]
New benchstat output:
```
$ benchstat old.bench new.bench 
name                 old time/op    new time/op    delta
StartRequestSpan-16    1.08µs ±13%    0.35µs ± 3%  -67.92%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
StartRequestSpan-16      768B ± 0%      264B ± 0%  -65.62%  (p=0.000 n=10+10)

name                 old allocs/op  new allocs/op  delta
StartRequestSpan-16      21.0 ± 0%      10.0 ± 0%  -52.38%  (p=0.000 n=10+10)
```

**Pull request merged**